# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/gamepad/GamepadServiceTest.h
# Commit: ec6a64b639a8
# Full Hash: ec6a64b639a82e957376c09a71aef7ce62b3a5aa
# Author: Chris Martin <cmartin@mozilla.com>
# Date: 2020-12-03 09:47:26
# Regressor Bug: 1657404
# File Overlap Count: 3
# Description:
#   Bug 1657404 - Implement a strongly-typed, service-dependent gamepad handle r=handyman,aklotz
#   
#   Currently, the gamepad code uses a uint32_t as a handle and does some trickery
#   with it by trying to create a unique ID and adding an offset to it for VR code.
#   
# ==============================================================================

diff -r 6792f28b99bd -r ec6a64b639a8 dom/gamepad/GamepadServiceTest.h
--- a/dom/gamepad/GamepadServiceTest.h	Wed Dec 02 22:36:35 2020 +0000
+++ b/dom/gamepad/GamepadServiceTest.h	Wed Dec 02 23:06:05 2020 +0000
@@ -9,6 +9,7 @@
 
 #include "mozilla/DOMEventTargetHelper.h"
 #include "mozilla/dom/GamepadBinding.h"
+#include "mozilla/dom/GamepadHandle.h"
 #include "mozilla/dom/TypedArray.h"
 #include "mozilla/WeakPtr.h"
 
@@ -40,22 +41,22 @@
       const nsAString& aID, GamepadMappingType aMapping, GamepadHand aHand,
       uint32_t aNumButtons, uint32_t aNumAxes, uint32_t aNumHaptics,
       uint32_t aNumLightIndicator, uint32_t aNumTouchEvents, ErrorResult& aRv);
-  void ReplyGamepadIndex(uint32_t aPromiseId, uint32_t aIndex);
+  void ReplyGamepadHandle(uint32_t aPromiseId, const GamepadHandle& aHandle);
 
-  void RemoveGamepad(uint32_t aIndex);
-  void NewButtonEvent(uint32_t aIndex, uint32_t aButton, bool aPressed,
+  void RemoveGamepad(uint32_t aHandleSlot);
+  void NewButtonEvent(uint32_t aHandleSlot, uint32_t aButton, bool aPressed,
                       bool aTouched);
-  void NewButtonValueEvent(uint32_t aIndex, uint32_t aButton, bool aPressed,
-                           bool aTouched, double aValue);
-  void NewAxisMoveEvent(uint32_t aIndex, uint32_t aAxis, double aValue);
-  void NewPoseMove(uint32_t aIndex, const Nullable<Float32Array>& aOrient,
+  void NewButtonValueEvent(uint32_t aHandleSlot, uint32_t aButton,
+                           bool aPressed, bool aTouched, double aValue);
+  void NewAxisMoveEvent(uint32_t aHandleSlot, uint32_t aAxis, double aValue);
+  void NewPoseMove(uint32_t aHandleSlot, const Nullable<Float32Array>& aOrient,
                    const Nullable<Float32Array>& aPos,
                    const Nullable<Float32Array>& aAngVelocity,
                    const Nullable<Float32Array>& aAngAcceleration,
                    const Nullable<Float32Array>& aLinVelocity,
                    const Nullable<Float32Array>& aLinAcceleration);
-  void NewTouch(uint32_t aIndex, uint32_t aTouchArrayIndex, uint32_t aTouchId,
-                uint8_t aSurfaceId, const Float32Array& aPos,
+  void NewTouch(uint32_t aHandleSlot, uint32_t aTouchArrayIndex,
+                uint32_t aTouchId, uint8_t aSurfaceId, const Float32Array& aPos,
                 const Nullable<Float32Array>& aSurfDim);
   void Shutdown();
 
@@ -76,6 +77,7 @@
   // will only be used in this singleton class and deleted during the IPDL
   // shutdown chain
   RefPtr<GamepadTestChannelChild> mChild;
+  nsTArray<GamepadHandle> mGamepadHandles;
 
   nsRefPtrHashtable<nsUint32HashKey, dom::Promise> mPromiseList;
 
@@ -83,6 +85,10 @@
   ~GamepadServiceTest();
   void InitPBackgroundActor();
   void DestroyPBackgroundActor();
+
+  uint32_t AddGamepadHandle(GamepadHandle aHandle);
+  void RemoveGamepadHandle(uint32_t aHandleSlot);
+  GamepadHandle GetHandleInSlot(uint32_t aHandleSlot) const;
 };
 
 }  // namespace dom